enterprise/server/remote_execution/containers/ociruntime/ociruntime.go

```diff
@@ -1293,58 +1294,97 @@ func downloadLayer(ctx context.Context, layer ctr.Layer, destDir string) error {
 	}
 	defer os.RemoveAll(tempUnpackDir)
 
-	// TODO: avoid tar command.
-	cmd := exec.CommandContext(ctx, "tar", "--extract", "--directory", tempUnpackDir)
-	var stderr bytes.Buffer
-	cmd.Stdin = rc
-	cmd.Stderr = &stderr
-	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
-	if err := cmd.Run(); err != nil {
-		return status.UnavailableErrorf("download and extract layer tarball: %s: %q", err, stderr.String())
-	}
-
-	// Convert whiteout files to overlayfs format.
-	err = filepath.WalkDir(tempUnpackDir, func(path string, entry fs.DirEntry, err error) error {
+	tr := tar.NewReader(rc)
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
 		if err != nil {
-			return err
+			return status.UnavailableErrorf("download and extract layer tarball: %s", err)
 		}
 
-		if !entry.Type().IsRegular() {
-			return nil
+		cleanHeaderName := filepath.Clean(header.Name)
+		if cleanHeaderName != header.Name {
+			names := strings.Split(header.Name, string(os.PathSeparator))
+			for _, n := range names {
+				if n == ".." {
+					// Don't suppport backtracking paths.
+					return status.UnavailableErrorf("tar entry is not clean: %q", header.Name)
+				}
+			}
 		}
```

Comment:
The `cleanHeaderName != header.Name` check seems undesired - `filepath.Clean("../foo")` returns `"../foo"` so the check would not be run.

```suggestion
		if slices.Contains(filepath.Split(header.Name), "..") {
			return status.UnavailableError("tar entry contains disallowed parent directory reference", header.Name)
		}
```

Reply:
Oh wat. We probably dont want filepath.Split, but i agree with the fix

Reply:
woops, filepath.Split doesn't do what I thought it did :skull: